### ISSUE TITLE
Persistent Object State

### ISSUE THREAD

**Issue by jollen** [w3c/web-of-things-framework#12]
Things objects could be initialized (registered) from a data store. (for example memory caching like memcached). The client can manage the IoT device states. For example, the previous states of backdoor light from a memcached is on. Thus, the WoT server could restart without reseting all of the device controlling states.

**Comment by draggett** [comment]
> On 5 Jun 2015, at 14:19, jollen <notifications@github.com> wrote:
> 
> Things objects could be initialized (registered) from a data store. (for example memory caching like memcached). The client can manage the IoT device states. For example, the previous states of backdoor light from a memcached is on. Thus, the WoT server could restart without reseting all of the device controlling states.
> 

Sounds like a good idea, and might also be useful for things that have been restarted.  In other cases, the thing state can be read from the IoT device.

—
   Dave Raggett <dsr@w3.org <mailto:dsr@w3.org>>

**Comment by mpetyx** [comment]
Is there any progress on this?
This could be easily tackled with something like a light nosql (in browser) database similar to http://lokijs.org/ .

**Comment by bubbafat** [comment]
I think there is a lot of value in a store of things - and also in persisting state during run-time - but it may be worthwhile to think through the scenarios being solved.

Loading the initial state could be orthogonal to the framework because the loader would only need access to the wot.thing/register_proxy APIs.  Basically it would replace what is in demo.js.  Saving, though, is another story.  I believe we should avoid binding the framework to a specific persistence layer - but rather provide the necessary hooks to make it possible for any layer to be used.

My gut feeling is that if we used something like EventEmitter to fire change events the persistence layer could subscribe to those and perform whatever actions it desires.  But how does the layer subscribe?

My thinking is that if we had a "Thing Created" event that fired immediately after (before?) the thing's start() function was called, that the persistence layer could subscribe to that to become aware of the thing and then subscribe to it's "Changed" event.  I think we would have the generated property fire the changed event.

This would mean that a new persistence layer would only need to handle that one event to minimally save updates.

We could go further, though.

* An event that resets the thing to it's previously saved state (which could eventually become the start of loosely transactional behaviors).
* Perhaps some properties should be attributed as not persisted.  This might be sensitive, temporary or calculated values that should not end up in a persistent data store.
* Lazy-loading things - when a request for a thing (e.g., a proxy) comes in, if the thing does not exist the framework could fire a "resolve" event that would allow the item to be loaded.

And at the risk of derailing this conversation... I also think that a discussion around persistence would be an appropriate time to talk about linked models because it could affect persistence.

Let's say we're Nest and have 1 million thermostats in the market.  Some are on consumer walls and some are on store shelves.  But all of them are pre-registered into a persistence layer (or even created on-the-fly using the resolve event).

There isn't a good reason that all 1 million entries need to have their own model.  They could just as well have a linked model - something like the JSON-LD context.  E.g.,

```JavaScript
{
    "@context":
    {
        "name": "Nest Model 123 fw version 2.38.2a",
        "@model": {
            "@properties": ...
            "@events": ...
            "@dependencies": ...
        },
        "@implementation": {
            ...
        }
    }
}
```

Then in the thing:

```JavaScript
{
    "@context": "http://our.url/contexts/Nest123-2-38-2a.jsonld"
    "name": "/wot/nest12",
    "@values": {
        ... the persisted values
    }    
}
```

**Comment by draggett** [comment]
Thanks for the analysis. I agree that there is potential value in having events that signal when a thing has started, and when a property is updated.  The latter arises due to the potential lag in the property being updated on a battery operated device that spends most of its time asleep, so that the update won't be received until the device's next receive window.  An event would avoid the need for the script to poll the property to determine when it has been updated. This would simplify the code for a user experience that provides a visible indication of the lag.

In respect to persistence, for sensor values you should be able to read the value from the sensor, and it is unclear whether the old value from when the system was last turned off is better than "undefined".  The location of a car is likely to remain the same when the car is turned off, but this isn't true for a mobile phone.  For properties that aren't tied to sensor values, persistence seems very reasonable. This would cover preference settings, and actuator values.

One idea is to add a metadata flag to the data model to indicate whether a property should be persisted, e.g. here is a property for a thermostat that has units, min and max values, and should be persisted when the system is shutdown:
```
   temperature: {
       type: number,
       min: 12,
       max: 25,
       units: "celsius",
       persist: true
   }
```

**Comment by bubbafat** [comment]
I agree - that is pretty much what I was thinking of in my second bullet point.

I like the car vs phone analogy.  I could see a car wanting to record it's "last known location" but that would be distinct from it's "current location" - this approach would allow them to have both properties while making it clearer which is meant to be used in each context.

Now we just need to figure out how setting thing.location causes an event to be fired that allows the thing to set thing.last_known_location.  

So I'll go off the rails a little bit here...

```JavaScript
temperature: {
    type: number,
    min: 12,
    max: 25,
    units: "celsius",
    persist: true,
    onChange: "set_last_known_temp"
}

last_known_temperature: {
    inherits: "temperature",  // just saying "use temperature's settings" - might be a neat thing, though
    persist: false
}

// then define the action 

set_last_known_temp: function(thing) {
    thing.last_known_temp = thing.temperature;
}
```

**Comment by draggett** [comment]
> we now need some mechanism for setting thing.location to update thing.last_known_location

You could write an event handler for that using the event the signals changes to the location property. However, my metadata idea seems simpler for developers by passing the buck for implementing persistence to a system programmer. This is a general principle of distributing tasks to people based upon their skill sets.
